kernel/src/collections/list_v.rs

```diff
@@ -0,0 +1,606 @@
+                        ==> #[trigger] ghost_state@.points_to_map.dom().contains(i)
+                        && ghost_state@.points_to_map[i].is_init()
+                        && ghost_state@.points_to_map[i].id() == ghost_state@.cells[i as int].id(),
+                forall|i: nat|
+                    self.index@ <= i < (ghost_state@.cells.len() - 1) as nat
+                        ==> match #[trigger] ghost_state@.points_to_map[i].value() {
+                        Option::Some(_) => true,
+                        Option::None => false,
+                    },
+                match ghost_state@.points_to_map[(ghost_state@.cells.len() - 1) as nat].value() {
+                    Option::Some(_) => false,
+                    Option::None => true,
+                },
+                self.cur == ghost_state@.points_to_map[self.index@].value(),
+                match last {
+                    Option::Some(_) => ghost_state@.points_to_map.dom().contains(
+                        (self.index@ - 1) as nat,
+                    ) && last == ghost_state@.points_to_map[(self.index@ - 1) as nat].value(),
+                    Option::None => self.index@ == old(self).index@,
+                },
+            ensures
+                ghost_state@.cells.len() == self.index@ + 1,
+                self.cur == ghost_state@.points_to_map[self.index@].value(),
+                match last {
+                    Option::Some(_) => ghost_state@.points_to_map.dom().contains(
+                        (self.index@ - 1) as nat,
+                    ) && last == ghost_state@.points_to_map[(self.index@ - 1) as nat].value(),
+                    Option::None => old(self).index@ + 1 == ghost_state@.cells.len(),
+                },
+        {
+            let cur = self.next(ghost_state);
+            if cur.is_none() {
+                proof {
+                    assert(match last {
+                        Option::Some(_) => ghost_state@.points_to_map.dom().contains(
+                            (self.index@ - 1) as nat,
+                        ) && last == ghost_state@.points_to_map[(self.index@ - 1) as nat].value(),
+                        Option::None => old(self).index@ + 1 == ghost_state@.cells.len(),
+                    });
+                }
+                break ;
+            } else {
+                proof {
+                    assert(match cur {
+                        Option::Some(_) => ghost_state@.points_to_map.dom().contains(
+                            (self.index@ - 1) as nat,
+                        ) && cur == ghost_state@.points_to_map[(self.index@ - 1) as nat].value(),
+                        Option::None => false,
+                    });
+                }
+                last = cur;
+            }
+        }
+        last
+    }
+}
+
+impl<'a, T: ?Sized + ListNodeV<'a, T>> ListV<'a, T> {
+    pub closed spec fn well_formed_list(&self, ghost_state: &Tracked<GhostState<'a, T>>) -> bool {
+        &&& ghost_state@.cells.len() >= 1
```

Comment:
Just to confirm, that does not mean the list cannot be empty right? The `ListLink` can still be empty?

Reply:
The `ListV` can be empty, but the `ghost_state@.cells` cannot because the last one in the sequence always contains `None` in the cell. So the actual length of `ListV` is `ghost_state@.cells.len() - 1`.